### DISCUSSION TITLE
SPI Configuration

### DISCUSSION THREAD

**Discussion by dasniko** [keycloak/keycloak#10311] [Archive]
Hi,

in the [community design doc there is a chapter](https://github.com/keycloak/keycloak-community/blob/main/design/keycloak.x/configuration.md#spis-and-provider-configuration) about **SPI** and **Provider** configuration. Do only I miss this completely in the new docs? Is it contained anywhere or is it really still missing and will be published soon?

---

Currently, I'm struggling a bit configuring a login provider.
In Wildfly, I used this CLI:

```
/subsystem=keycloak-server/spi=login:add(default-provider=myAwesomeFreemarkerProvider)
/subsystem=keycloak-server/spi=login/provider=myAwesomeFreemarkerProvider(enabled=true)
```

How will I do this with the `KC_SPI_LOGIN_...` environment variables (or `spi-login-...` properties)?
Simply this?
```
KC_SPI_LOGIN_DEFAULT_PROVIDER=myAwesomeFreemarkerProvider
KC_SPI_LOGIN_MY_AWESOME_FREEMARKER_PROVIDER_ENABLED=true
```

@DGuhr @pedroigor ?

**Comment by thomasdarimont** [comment]
Setting a custom login-forms provider works for me with:
```
--spi-login-provider=acme-custom-login-forms
```

```
KC_SPI_LOGIN_PROVIDER=acme-custom-login-forms
```

However once the server updated the server configuration after `build` / `--auto-build` one cannot switch the
login-provider back anymore...

I think support for something like `--spi-login-provider=default` would be helpful here, to revert the override.

**Reply by dasniko** [reply]
Thanks, @thomasdarimont that works for me.

Although, a detailed and official guide would be helpful and would make things more clear for the migration path!

**Reply by DGuhr** [reply]
@thomasdarimont could you please create an issue around the switching behavior so this doesn't get lost? thanks! ;)

For the guide part: Think I'll work in that area next week.

**Reply by thomasdarimont** [reply]
@DGuhr I created issue https://github.com/keycloak/keycloak/issues/10327

**Reply by DGuhr** [reply]
btw: if anyone wants to take the time and write a standalone.xml->keycloak.conf parser, that would be more than nice. I will most likely not have the time to do it in the foreseeable future, but can surely help / provide guidance. So when somebody searches a valuable pet project in the keycloak ecosystem, feel free to reach out.😇

**Reply by thomasdarimont** [reply]
What exactly should this tool do?
I assume that you want to extract non-default keycloak-server config and transform the config to the new format.

Should this be a separate shell tool? 
Perhaps a simple jbang script will do the job here.
(It's probably just for wildfly or also jboss eap)?

**Reply by DGuhr** [reply]
Right. Would be nice if you could just do `kc.sh tools migrate-config --file=<your-legacy-standalone.xml>` and you get a valid keycloak.conf file containing all your customizations out of it, wouldn't it? Best would be wildfly and eap, for sure.

**Reply by stianst** [reply]
I'd love to be proven wrong here, but I have my doubts that it is realistic to build a tool that can migrate standalone.xml in a meaningful way.

**Reply by pedroigor** [reply]
@thomasdarimont @dasniko Could you please check if this new guide helps you? See https://github.com/keycloak/keycloak/pull/10401.

**Comment by mduchrow** [comment]
Well, one of my customers has custom providers with configurations like this:

> <spi name="authenticator">
	<provider name="custom-username-password-form" enabled="true">
		<properties>
			<property name="selfRegistrationURL" value="http://example.com/register"/>
		</properties>
	</provider>
</spi>

How am I supposed to configure this in the new Keycloak style?

The following attempts all failed:

* KC_SPI_CUSTOM_USERNAME_PASSWORD_FORM_selfRegistrationURL=http://example.com/register
* KC_SPI_CUSTOM_USERNAME_PASSWORD_FORM_SELFREGISTRATIONURL=http://example.com/register
* KC_SPI_CUSTOM-USERNAME-PASSWORD-FORM_selfRegistrationURL=http://example.com/register
* KC_SPI_CUSTOM-USERNAME-PASSWORD-FORM_SELFREGISTRATIONURL=http://example.com/register

We can change the name of the property, but what's about the provider ID? Changing it is not possible.

**Reply by dasniko** [reply]
I'd assume from the docs (#10401) it will be
`spi-custom-username-password-form-self-registration-url=http://example.com/register`
and 
`KC_SPI_CUSTOM_USERNAME_PASSWORD_FORM_SELF_REGISTRATION_URL=http://example.com/register`

`selfRegistrationURL` is camel-case and thus would change to `-` as separator and lower case, for envvars all is uppercase and `-` turns into `_`.
But I'm not sure about the `URL` part, will it stay `url` or will it become `u-r-l`? 🤔

**Reply by mduchrow** [reply]
Thanks Niko, but that didn't work.
I also tried with property renamed to "self-registration-url" and "selfregistrationurl".
No success.
I wonder how Keycloak recognizes where the provider name ends and the property name starts if everything is filled with dashes.

**Reply by mduchrow** [reply]
What eventually was successful:
KC_SPI_AUTHENTICATOR_CUSTOM_USERNAME_PASSWORD_FORM_SELFREGISTRATIONURL=http://example.com/register
for property name "selfregistrationurl" used inside the factory init method.
And also
KC_SPI_AUTHENTICATOR_CUSTOM_USERNAME_PASSWORD_FORM_SELF_REGISTRATION_URL=http://example.com/register
for property name "self-registration-url".

**Reply by dasniko** [reply]
Ok, from your example it wasn't obvious that it's an authenticator SPI and I completely missed the part of the SPI-ID.

**Reply by mduchrow** [reply]
Because it wasn't obvious from the original example that the spi was "authenticator", here's the summary how it works:
Old Wildfly style configuration in standalone.xml:
````xml
<spi name="authenticator">
    <provider name="custom-username-password-form" enabled="true">
        <properties>
            <property name="selfRegistrationURL" value="http://example.com/register"/>
        </properties>
    </provider>
</spi>
````
New style configuration:
**KC_SPI_AUTHENTICATOR_CUSTOM_USERNAME_PASSWORD_FORM_SELF_REGISTRATION_URL**=http://example.com/register

This even works with camel-case property name "selfRegistrationURL"!

**Comment by stianst** [comment]
The formats for the SPI properties are really fugly for sure. It also doesn't help that built-in SPIs and providers have a mix of naming schemes (cammel-case or '-' separated).

@DGuhr @pedroigor this is really something we should improve on at some point
